core/src/Kokkos_GraphNode.hpp

```diff
@@ -0,0 +1,469 @@
+                         Functor&& functor) const {
+    // needs to static assert constraint: DataParallelFunctor<Functor>
+    return this->then_parallel_for(std::move(name),
+                                   Kokkos::RangePolicy<execution_space>(0, n),
+                                   (Functor &&) functor);
+  }
+
+  template <class Functor>
+  auto then_parallel_for(std::size_t n, Functor&& functor) const {
+    // needs to static assert constraint: DataParallelFunctor<Functor>
+    return this->then_parallel_for("", n, (Functor &&) functor);
+  }
+
+  // </editor-fold> end then_parallel_for }}}2
+  //----------------------------------------------------------------------------
+
+  //----------------------------------------------------------------------------
+  // <editor-fold desc="then_parallel_reduce"> {{{2
+
+  template <
+      class Policy, class Functor, class ReturnType,
+      typename std::enable_if<
+          // equivalent to:
+          //   requires Kokkos::ExecutionPolicy<remove_cvref_t<Policy>>
+          is_execution_policy<Kokkos::Impl::remove_cvref_t<Policy>>::value,
+          // --------------------
+          int>::type = 0>
+  auto then_parallel_reduce(std::string arg_name, Policy&& arg_policy,
+                            Functor&& functor,
+                            ReturnType&& return_value) const {
+    auto graph_impl_ptr = m_graph_impl.lock();
+    KOKKOS_EXPECTS(bool(graph_impl_ptr))
+    KOKKOS_EXPECTS(bool(m_node_impl))
+    // TODO @graph restore this expectation once we add comparability to space
+    //      instances
+    // KOKKOS_EXPECTS(
+    //   arg_policy.space() == m_graph_impl->get_execution_space());
+
+    // needs static assertion of constraint:
+    //   DataParallelReductionFunctor<Functor, ReturnType>
+
+    using policy_t = typename std::remove_cv<
+        typename std::remove_reference<Policy>::type>::type;
+    static_assert(
+        std::is_same<typename policy_t::execution_space,
+                     execution_space>::value,
+        // TODO @graph make defaulted execution space work
+        // || policy_t::execution_space_is_defaulted,
+        "Execution Space mismatch between execution policy and graph");
+
+    // This is also just an expectation, but it's one that we expect the user
+    // to interact with (even in release mode), so we should throw an exception
+    // with an explanation rather than just doing a contract assertion.
+    if (Kokkos::Impl::parallel_reduce_needs_fence(
+            graph_impl_ptr->get_execution_space(), return_value)) {
+      Kokkos::Impl::throw_runtime_exception(
+          "Parallel reductions in graphs can't operate on Reducers that "
+          "reference a scalar because they can't complete synchronously. Use a "
+          "Kokkos::View instead and keep in mind the result will only be "
+          "available once the graph is submitted (or in tasks that depend on "
+          "this one).");
```

Comment:
Yeah, because of the way Reducers are set up to use a boolean (rather than a template parameter, say) to determine whether the pointer held by the Reducer is a View or not. Not my preference, but it's what we're stuck with.

Reply:
Resolved in 0ffe95f